### PR TITLE
Horovod

### PR DESCRIPTION
**Pull request by icywang86rui**
*Issue #, if available:*

*Description of changes:*

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

- [ ] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [ ] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [ ] I have updated the [changelog](https://github.com/aws/sagemaker-python-sdk/blob/master/CHANGELOG.rst) with a description of my changes (if appropriate)
- [ ] I have updated any necessary [documentation](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/data/tensorflow_mnist/horovod_mnist.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+learn = tf.contrib.learn
+
+tf.logging.set_verbosity(tf.logging.INFO)
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser()
+    # Data, model, and output directories
+    parser.add_argument('--output-data-dir', type=str, default=os.environ['SM_OUTPUT_DATA_DIR'])
+    parser.add_argument('--model_dir', type=str, default=os.environ['SM_MODEL_DIR'])
```
Comment (mvsusp) [inline tests/data/tensorflow_mnist/horovod_mnist.py]: Let's remove the default here given that is passed through the hyperparameters.
Use os.environ.get instead to avoid errors running the script outside SageMaker.

File: tests/data/tensorflow_mnist/horovod_mnist.py [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+
+
+def main(_):
+    args, unknown = _parse_args()
+
+    # Horovod: initialize Horovod.
+    hvd.init()
+
+    # Download and load MNIST dataset.
+    mnist = learn.datasets.mnist.read_data_sets('MNIST-data-%d' % hvd.rank())
```
Comment (mvsusp) [inline tests/data/tensorflow_mnist/horovod_mnist.py]: What is the size of the dataset?

Comment (icywang86rui) [reply tests/data/tensorflow_mnist/horovod_mnist.py]: the training data is 164M. With the eval data and the label it's about 200M.

File: tests/data/tensorflow_mnist/horovod_mnist.py (hunk 1 above)
Comment (icywang86rui) [reply tests/data/tensorflow_mnist/horovod_mnist.py]: k

File: src/sagemaker/tensorflow/README.rst [hunk 3]
```diff
@@ -207,6 +207,43 @@ After attaching, the estimator can be deployed as usual.
...
+
+  from sagemaker.tensorflow import TensorFlow
+
+  tf_estimator = TensorFlow(entry_point='tf-train.py', role='SageMakerRole',
+                            train_instance_count=2, train_instance_type='ml.p2.xlarge',
+                            framework_version='1.11', py_version='py3',
+                            distributions={'mpi': {'enabled': True}})
+  tf_estimator.fit('s3://bucket/path/to/training/data')
+
+If MPI is enabled the container will construct and run MPI commands which executes your training script. You can find
```
[the remark this answers is not in the document]
Comment (icywang86rui) [reply src/sagemaker/tensorflow/README.rst]: k

File: src/sagemaker/tensorflow/README.rst [hunk 4]
```diff
@@ -207,6 +207,43 @@ After attaching, the estimator can be deployed as usual.
 
     tf_estimator = TensorFlow.attach(training_job_name=training_job_name)
 
+Distributed Training
+''''''''''''''''''''
+
+To run your training job in a distributed fashion you need to set ``train_instance_count`` to a number larger than 1.
+We support two different types of distributed training, parameter server and MPI. The ``distributions`` parameter is
```
[the remark this answers is not in the document]
Comment (icywang86rui) [reply src/sagemaker/tensorflow/README.rst]: I think it's possible for user to run other types of distributed training. But I wouldn't say those are supported. These two types are setup by our code and we are going to support and maintain that code.

File: src/sagemaker/tensorflow/README.rst [hunk 5]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
 
     tf_estimator = TensorFlow.attach(training_job_name=training_job_name)
 
+Distributed Training
+''''''''''''''''''''
+
+To run your training job with multiple instances in a distributed fashion you need to set ``train_instance_count``
+to a number larger than 1. We support two different types of distributed training, parameter server and Horovod.
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/README.rst]: is running a script that uses MPI but not Horovod a use case?

File: src/sagemaker/tensorflow/README.rst [hunk 6]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
...
+"""""""""""""""""""""
+
+Horovod is a distributed training framework based on MPI. You can find more details `here <https://github.com/uber/horovod>`__.
+
+The container sets up the MPI environment and executes the ``mpirun`` command enabling you to run any Horovod
+training script with Script Mode.
+
+Training with ``MPI`` is configured by specifying following fields in ``distributions``:
+
+- ``enabled (bool)``: If set to `True`, the MPI setup is performed and ``mpirun`` command is executed.
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/README.rst]: double backticks for `True`

File: src/sagemaker/tensorflow/README.rst [hunk 7]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
...
+In the below example we create an estimator to launch Horovod distributed training with 2 processes on one host:
+
+.. code:: python
+
+    from sagemaker.tensorflow import TensorFlow
+
+    tf_estimator = TensorFlow(entry_point='tf-train.py', role='SageMakerRole',
+                            train_instance_count=1, train_instance_type='ml.p2.xlarge',
+                            framework_version='1.11', py_version='py3',
+                            distributions: {
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/README.rst]: line up the arguments, and also s/: /=

File: src/sagemaker/tensorflow/README.rst (hunk 5 above)
Comment (icywang86rui) [reply src/sagemaker/tensorflow/README.rst]: I think it is but if you use the tensorflow container it uses Horovod. I could be wrong.

File: src/sagemaker/tensorflow/README.rst [hunk 8]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
...
+    from sagemaker.tensorflow import TensorFlow
+
+    tf_estimator = TensorFlow(entry_point='tf-train.py', role='SageMakerRole',
+                            train_instance_count=1, train_instance_type='ml.p2.xlarge',
+                            framework_version='1.11', py_version='py3',
+                            distributions={
+                                "mpi":{
+                                    "enabled":True,
+                                    "processes_per_host":2,
+                                    "custom_mpi_options": "--NCCL_DEBUG INFO"
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/README.rst]: single quotes for strings

File: src/sagemaker/tensorflow/README.rst [hunk 9]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
...
+
+    from sagemaker.tensorflow import TensorFlow
+
+    tf_estimator = TensorFlow(entry_point='tf-train.py', role='SageMakerRole',
+                            train_instance_count=1, train_instance_type='ml.p2.xlarge',
+                            framework_version='1.11', py_version='py3',
+                            distributions={
+                                "mpi":{
+                                    "enabled":True,
+                                    "processes_per_host":2,
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/README.rst]: spaces after the colons

File: src/sagemaker/tensorflow/estimator.py [hunk 10]
```diff
@@ -199,15 +199,22 @@ def __init__(self, training_steps=None, evaluation_steps=None, checkpoint_path=N
                         custom-image:latest.
             script_mode (bool): If set to True will the estimator will use the Script Mode containers (default: False).
                 This will be ignored if py_version is set to 'py3'.
-            distribution (dict): A dictionary with information on how to run distributed training
-                (default: None). Currently we only support distributed training with parameter servers. To enable it
-                use the following setup:
+            distributions (dict): A dictionary with information on how to run distributed training
+                (default: None). Currently we support distributed training with parameter servers and MPI. To enable
+                parameter server use the following setup:
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/estimator.py]: s/server/servers

File: src/sagemaker/tensorflow/estimator.py [hunk 11]
```diff
@@ -435,10 +453,12 @@ def hyperparameters(self):
         hyperparameters.update(Framework._json_encode_hyperparameters(additional_hyperparameters))
         return hyperparameters
 
-    def _default_s3_path(self, directory):
+    def _default_s3_path(self, directory, mpi=False):
         local_code = get_config_value('local.local_code', self.sagemaker_session.config)
         if self.sagemaker_session.local_mode and local_code:
             return '/opt/ml/shared/{}'.format(directory)
+        elif mpi:
+            return '/opt/ml/model'
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/estimator.py]: should we make this a constant?

File: tests/integ/test_tf_script_mode.py [hunk 12]
```diff
@@ -74,6 +75,28 @@ def test_mnist_distributed(sagemaker_session, instance_type):
                            ['graph.pbtxt', 'model.ckpt-0.index', 'model.ckpt-0.meta', 'saved_model.pb'])
 
 
+@pytest.mark.skipif(integ.PYTHON_VERSION != 'py3', reason="Script Mode tests are only configured to run with Python 3")
```
Comment (laurenyu) [inline tests/integ/test_tf_script_mode.py]: single quotes for strings

File: src/sagemaker/tensorflow/README.rst (hunk 8 above)
Comment (yangaws) [reply src/sagemaker/tensorflow/README.rst]: Updated.

File: src/sagemaker/tensorflow/README.rst [hunk 13]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
 
     tf_estimator = TensorFlow.attach(training_job_name=training_job_name)
 
+Distributed Training
+''''''''''''''''''''
+
+To run your training job with multiple instances in a distributed fashion you need to set ``train_instance_count``
```
[the remark this answers is not in the document]
Comment (yangaws) [reply src/sagemaker/tensorflow/README.rst]: Updated

File: src/sagemaker/tensorflow/README.rst [hunk 14]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
...
+to a number larger than 1. We support two different types of distributed training, parameter server and Horovod.
+The ``distributions`` parameter is used to configure which distributed training strategy to use.
+
+Training with parameter servers
+"""""""""""""""""""""""""""""""
+
+If parameter server is enabled, the container will launch a parameter server thread in each instance first then execute
```
[the remark this answers is not in the document]
Comment (yangaws) [reply src/sagemaker/tensorflow/README.rst]: Updated

File: src/sagemaker/tensorflow/README.rst [hunk 15]
```diff
@@ -207,6 +207,63 @@ After attaching, the estimator can be deployed as usual.
...
+  tf_estimator = TensorFlow(entry_point='tf-train.py', role='SageMakerRole',
+                            train_instance_count=2, train_instance_type='ml.p2.xlarge',
+                            framework_version='1.11', py_version='py3',
+                            distributions={'parameter_server': {'enabled': True}})
+  tf_estimator.fit('s3://bucket/path/to/training/data')
+
+Training with Horovod
+"""""""""""""""""""""
+
+Horovod is a distributed training framework based on MPI. You can find more details in `Horovod README <https://github.com/uber/horovod>`__.
```
[the remark this answers is not in the document]
Comment (yangaws) [reply src/sagemaker/tensorflow/README.rst]: Updated

File: src/sagemaker/tensorflow/README.rst [hunk 16]
```diff
@@ -207,6 +207,64 @@ After attaching, the estimator can be deployed as usual.
...
+''''''''''''''''''''
+
+To run your training job with multiple instances in a distributed fashion, set ``train_instance_count``
+to a number larger than 1. We support two different types of distributed training, parameter server and Horovod.
+The ``distributions`` parameter is used to configure which distributed training strategy to use.
+
+Training with parameter servers
+"""""""""""""""""""""""""""""""
+
+If you specify parameter_server as the value of the distributions parameter, the container launches a parameter server
```
Comment (laurenyu) [inline src/sagemaker/tensorflow/README.rst]: backticks around `parameter_server`

File: src/sagemaker/tensorflow/estimator.py (hunk 10 above)
Comment (laurenyu) [reply src/sagemaker/tensorflow/estimator.py]: for "To enable parameter server" - s/server/servers

File: tests/integ/test_tf_script_mode.py (hunk 12 above)
Comment (laurenyu) [reply tests/integ/test_tf_script_mode.py]: single quotes for the reason string

File: tests/data/tensorflow_mnist/horovod_mnist.py [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+learn = tf.contrib.learn
+
+tf.logging.set_verbosity(tf.logging.INFO)
+
+
+def _parse_args():
+    parser = argparse.ArgumentParser()
+    # Data, model, and output directories
+    parser.add_argument('--output-data-dir', type=str, default=os.environ.get('SM_OUTPUT_DATA_DIR'))
+    parser.add_argument('--model_dir', type=str)
```
Comment (laurenyu) [inline tests/data/tensorflow_mnist/horovod_mnist.py]: nit: it's strange to me that we would mix underscores and hyphens in our examples like this

File: src/sagemaker/tensorflow/README.rst (hunk 5 above)
Comment (mvsusp) [reply src/sagemaker/tensorflow/README.rst]: Yes, MPI without Horovod is a valid use case.

Comment (laurenyu) [reply src/sagemaker/tensorflow/README.rst]: if that's the case, then I think this should be changed to something like "We support two different ways of handling distributed training: parameter servers and MPI. The use of MPI can be with or without Horovod." maybe include a link to Horovod documentation as well.
